File: packages/type-system/rust/build-wasm.py
import subprocess
import json
from pathlib import Path
from typing import Dict
import logging


def run_wasm_pack(target, out_dir: Path):
    logging.debug("Running wasm-pack, targeting %s, for %s", target, out_dir)
    subprocess.run(["wasm-pack", "build", "--target", target, "--out-dir", str(out_dir), "--out-name", "index", "--scope", "blockprotocol"])


def patch_package_json(file: Path, unique_patch: Dict):
    logging.debug("Patching %s", str(file))

    patch = {
        "homepage": "https://blockprotocol.org",
        "repository": {
            "type": "git",
            "url": "git@github.com:blockprotocol/blockprotocol.git",
            "directory": "packages/react-block-loader"
        },
        "license": "MIT",
        "author": {
            "name": "HASH",
            "url": "https://hash.ai"
        },
    }
    patch.update(unique_patch)

    with file.open("r+") as fp:
        contents = json.load(fp)
        fp.seek(0)
        contents.update(patch)
        json.dump(contents, fp)


def run_prettier(path: Path):
    logging.debug("Running prettier on %s", path)
    subprocess.run(["yarn", "prettier", "--write", str(path)])


def build_package(path: Path, target: str, patch: Dict):
    run_wasm_pack(target, path)
    package_json = (path / "package.json").resolve(True)
    patch_package_json(package_json, patch)
    run_prettier(package_json)


def main():
    web_patch = {"name": "@blockprotocol/type-system-web"}
    build_package(Path("../ts/type-system-web"), "web", web_patch)

    node_patch = {"name": "@blockprotocol/type-system-node"}
    build_package(Path("../ts/type-system-node"), "nodejs", node_patch)


if __name__ == "__main__":
    logging.basicConfig(level=logging.DEBUG)
    main()

File: packages/type-system/rust/build-wasm.ts
import * as path from "node:path";
import * as child_process from "child_process";
import * as fs from "fs";

// The PathObject definition in Node's path package
type PathObject = {
  dir: string;
  root: string;
  base: string;
  name: string;
  ext: string;
};

const runWasmPack = (target: string, outDir: PathObject) => {
  console.log(
    `Running wasm-pack, targeting ${target}, for ${path.format(outDir)}`,
  );
  const { error } = child_process.spawnSync("wasm-pack", [
    "build",
    "--target",
    target,
    "--out-dir",
    path.format(outDir),
    "--out-name",
    "index",
    "--scope",
    "blockprotocol",
  ]);
  if (error) {
    throw error;
  }
};

const patchPackageJson = (filePath: string, uniquePatch: object) => {
  console.log(`Patching ${filePath}`);

  let patch = {
    homepage: "https://blockprotocol.org",
    repository: {
      type: "git",
      url: "git@github.com:blockprotocol/blockprotocol.git",
    },
    license: "MIT",
    author: {
      name: "HASH",
      url: "https://hash.ai",
    },
    ...uniquePatch,
  };

  const contents = JSON.parse(fs.readFileSync(filePath, "utf8"));
  fs.writeFileSync(filePath, JSON.stringify({ ...contents, ...patch }));
};

const runPrettier = (filePath: string) => {
  const { error } = child_process.spawnSync("yarn", [
    "prettier",
    "--write",
    filePath,
  ]);
  if (error) {
    throw error;
  }
};

const buildWasmPackage = (
  packagePath: PathObject,
  target: string,
  patch: object,
) => {
  runWasmPack(target, packagePath);
  const packageJsonPath = path.resolve(
    path.join(path.format(packagePath), "package.json"),
  );
  patchPackageJson(packageJsonPath, patch);
  runPrettier(packageJsonPath);
};

const buildPackages = () => {
  const webPatch = { name: "@blockprotocol/type-system-web" };
  buildWasmPackage(path.parse("../ts/type-system-web"), "web", webPatch);

  const nodePatch = { name: "@blockprotocol/type-system-node" };
  buildWasmPackage(path.parse("../ts/type-system-node"), "nodejs", nodePatch);
};

buildPackages();
